src/index/mod.rs

```diff
@@ -467,6 +467,19 @@ impl TryFromInnerPtr for IndexImpl {
     }
 }
 
+pub trait UpcastIndex: NativeIndex {
```

Comment:
A brief documentation here would be nice. Presenting the use case that we know about here (keeping a `Vec<IndexImpl>`) might even help users with the same problem.